### PR TITLE
Feature/issue 2561 unit tests against cbs

### PR DESCRIPTION
**Pull request by tleyden**
Fixes https://github.com/couchbase/sync_gateway/issues/2561

Changes:

1. Consolidates test bucket creation
1. Easily switch test between running against Walrus vs Couchbase Server via environment variable
1. Set an env variable to run in `Xattr() mode`, which will skip certain tests if false
1. Moves bucket creation code out of test packages so can be referenced by other packages
1. Uncomments many previously commented tests
1. Skip tests that are known to be failing

PR notes:

- [ ]  Dependent SG-accel PR: https://github.com/couchbaselabs/sync-gateway-accel/pull/141
- [x] [Unit / integration tests passing against Couchbase Server 4.x with XATTRS=false](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/78/parameters/)
- [x] [Unit / integration tests passing against Couchbase Server 5.x with XATTRS=true](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/77/parameters/)
- [x] [Unit tests passing against walrus](http://uberjenkins.sc.couchbase.com/view/Build/job/build-sync-gateway/79/parameters/)

### CODE REVIEW

**Comment by tleyden** [comment]
Todo:


- [x] Reenable Xattr() tests if the Xattr env variable is true
- [x] EmptyTestBucket/CreateTestBucket -- tighten up sleep loop
- [x] enable the Couchbase-prefixed tests in index_writer_test.go
- [x] Figure out why TestSetBulk() not passing.	// TODO: why does this not returning an error casStale, err := bucket.WriteCas(key, 0, 0, casZero, "key-initial", sgbucket.Raw)

**Comment by tleyden** [comment]
@adamcfraser I tried this:

> enable the Couchbase-prefixed tests in index_writer_test.go

But I'm getting errors when running against a Couchbase Server bucket:

```
017-05-30T15:23:28.302-07:00 GoCB Opening Couchbase database test_indexbucket on <http://localhost:8091> as user "test_indexbucket"
2017-05-30T15:23:28.304-07:00 GoCB Opening Couchbase database test_indexbucket on <http://localhost:8091> as user "test_indexbucket"
2017-05-30T15:23:28.307-07:00 GoCB Opening Couchbase database test_indexbucket on <http://localhost:8091> as user "test_indexbucket"
2017-05-30 15:23:28.356040 I | channel set:{ABC}
2017-05-30T15:23:28.358-07:00 WARNING: channelComputer.ComputeChannelsForPrincipal failed on &{{naomi ABC:1  0  <nil>} { false [] <nil>   []} 0xc420523720 []}: not_found - Design document _design/sync_gateway not found -- auth.(*Authenticator).rebuildChannels() at auth.go:156
--- FAIL: TestIndexChangesAdminBackfill (9.66s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x108 pc=0x9acc2]

goroutine 7 [running]:
panic(0x7e2520, 0xc4200140f0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc420094780)
	/usr/local/go/src/testing/testing.go:579 +0x25d
panic(0x7e2520, 0xc4200140f0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/couchbaselabs/sync-gateway-accel/indexwriter.testUserDB(0xc4203c2b40, 0x8bcec7, 0x5, 0x8bf5d3, 0x7, 0xc4202afb40, 0x1, 0x1, 0x0)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/index_writer_test.go:364 +0x1b2
github.com/couchbaselabs/sync-gateway-accel/indexwriter.TestIndexChangesAdminBackfill(0xc420094780)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbaselabs/sync-gateway-accel/indexwriter/index_writer_test.go:62 +0x5cb
testing.tRunner(0xc420094780, 0x99c398)
	/usr/local/go/src/testing/testing.go:610 +0x81
```

So going to skip those tests for now.  (Will rename from Couchbase* -> Test*, but just call .skip())

**Comment by tleyden** [comment]
> Figure out why TestSetBulk() not passing.

@adamcfraser I spent a decent chunk of time on Friday trying to get to the bottom of it, but wasn't able to.  Since this test has not been running on a regular basis anyway (it's `CouchbaseTestSetBulk ` on the master branch), what about just spinning off a separate ticket to investigate to make sure it's not a real bug?

File: test_integration.sh [hunk 1]
```diff
@@ -0,0 +1,282 @@
+
+# This is a test suite to run the SG and SG Accel unit tests against a live Couchbase Server
+# bucket (as opposed to walrus).  It works around the "interference" issues that seem to happen
+# when trying to run the full test suite (eg, running test.sh) as follows (maybe due to residual objects in Sync Gateway
+# memory like timer callbacks)
+#
+# Regarding maintenance, this might be a temporary script until the interference issues mentioned above
+# are solved.  However, if it does become a permanent script, then it should probably be upgraded to
+# leverage a tool like https://github.com/ungerik/pkgreflect to discover the list of tests.
```
Comment (adamcfraser) [inline test_integration.sh:9]: Do we have a ticket filed for the interference issues, with some details about what exactly is going on?  In addition to the maintenance issues, I expect it also takes longer to run in this mode?

File: rest/utilities_testing.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+	if partyTime {
+		chans = channels.AtSequence(base.SetOf("*"), 1)
+	}
+	guest.SetExplicitChannels(chans)
+	a.Save(guest)
+}
+
+func (rt *RestTester) Close() {
+	if rt.RestTesterServerContext != nil {
+		rt.RestTesterServerContext.Close()
```
Comment (adamcfraser) [inline rest/utilities_testing.go:160]: Maybe worth noting regarding the test interference - I think that ServerContext.Close() includes some async shutdown processing (like the feed runner goroutine) that might be a problem.

File: auth/auth_test.go [hunk 3]
```diff
@@ -22,23 +22,6 @@ import (
...
-		Server:          base.UnitTestUrl(),
-		CouchbaseDriver: base.ChooseCouchbaseDriver(base.DataBucket),
-		BucketName:      "sync_gateway_tests"}, nil)
-	if err != nil {
-		log.Fatalf("Couldn't connect to bucket: %v", err)
-	}
-	if err != nil {
-		log.Fatalf("Couldn't install design doc: %v", err)
-	}
-}
```
Comment (adamcfraser) [inline auth/auth_test.go:40]: I'm very happy to see this code get pulled out - I've had it cause noise in a few different scenarios.

**Review by adamcfraser — APPROVED** [review]

File: test_integration.sh (hunk 1 above)
Comment (tleyden) [reply test_integration.sh:9]: Filed https://github.com/couchbase/sync_gateway/issues/2612 for this

File: rest/utilities_testing.go (hunk 2 above)
Comment (tleyden) [reply rest/utilities_testing.go:160]: Noted in #2612

**Pull request merged**